### PR TITLE
Refactor toolkit survey pop up for generic usage

### PR DESCRIPTION
**Pull request by Flanker32**
- Add entities for customer survey
- Refactor survey pop up for generic usage and move it to intellij library

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        @Override
+        public String getLink() {
+            return String.format(super.getLink(), getInstallationId());
+        }
+    },
+    AZURE_MGMT_SDK(Constants.AZURE_MGMT_SDK_SURVEY_LINK, Constants.AZURE_MGMT_SDK_SURVEY_DESCRIPTION, Constants.AZURE_ICON) {
+        @Override
+        public String getLink() {
+            return String.format(super.getLink(), getInstallationId());
+        }
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java:40]: how about making the getLink() implementation in their super interface as default ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        private static final Icon AZURE_ICON = IconLoader.getIcon("/icons/azure_large.png", CustomerSurvey.class);
+        // Survey for azure client&mgmt SDK users
+        private static final String AZURE_SDK_SURVEY_LINK = "https://www.surveymonkey.com/r/2D9YTRQ?src=intellij&ver=%s";
+        private static final String AZURE_SDK_SURVEY_DESCRIPTION = "Enjoy Azure SDKs?";
+        // Survey for azure client SDK users
+        private static final String AZURE_CLIENT_SDK_SURVEY_LINK = "https://www.surveymonkey.com/r/8P768ZY?src=intellij&ver=%s";
+        private static final String AZURE_CLIENT_SDK_SURVEY_DESCRIPTION = "Enjoy Azure SDKs?";
+        // Survey for azure mgmt SDK users
+        private static final String AZURE_MGMT_SDK_SURVEY_LINK = "https://www.surveymonkey.com/r/8HXDGTG?src=intellij&ver=%s";
+        private static final String AZURE_MGMT_SDK_SURVEY_DESCRIPTION = "Enjoy Azure SDKs?";
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java:79]: save values, different variables ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/SurveyPopUpDialog.java [hunk 3]
```diff
@@ -195,25 +206,12 @@ public void mouseDragged(MouseEvent e) {
...
-        if (!isDisposed) {
-            customerSurveyHelper.neverShowAgain();
-            close();
+    private synchronized void takeSurvey(CustomerSurveyResponse response) {
+        if (isDisposed) {
+            return;
         }
+        close();
+        this.listener.consume(response);
     }
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/SurveyPopUpDialog.java:215]: I recommend:

private void takeSurvey(...) {
     if (isDisposed) {
      return;
    }
    sync... { other codes }
}

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    AZURE_INTELLIJ_TOOLKIT(Constants.INTELLIJ_TOOLKIT_SURVEY_LINK, Constants.INTELLIJ_TOOLKIT_SURVEY_DESCRIPTION, Constants.AZURE_ICON) {
+        @Override
+        public String getLink() {
+            final String toolkit = Optional.ofNullable(PluginManagerCore.getPlugin(PluginId.getId("com.microsoft.tooling.msservices.intellij.azure")))
+                    .map(PluginDescriptor::getVersion).orElse(StringUtils.EMPTY);
+            final ApplicationInfo applicationInfo = ApplicationInfo.getInstance();
+            final String ide = String.format("%s %s", applicationInfo.getFullVersion(), applicationInfo.getBuild());
+            final String os = System.getProperty("os.name");
+            final String jdk = String.format("%s %s", System.getProperty("java.vendor"), System.getProperty("java.version"));
+            return String.format(super.getLink(), toolkit, ide, os, jdk, getInstallationId());
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java:52]: Is it possible to let the output of super.getLink() as a constant ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    private final Icon icon;
+
+    private static String getInstallationId() {
+        return AzureTelemeter.getCommonProperties().get("Installation ID");
+    }
+
+    @Override
+    public String getType() {
+        return this.name();
+    }
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java:67]: type is same with name ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/SurveyPopUpDialog.java [hunk 6]
```diff
@@ -47,52 +50,57 @@
...
     private JLabel lblClose;
 
-    private CustomerSurveyHelper customerSurveyHelper;
-
     private Point dragPosition;
-    private Timer disposeTimer;
-    private LafManagerListener themeListener;
-    private Color buttonOnHoverColor = Color.WHITE;
+    private Color buttonOnHoverColor = JBColor.WHITE;
     private boolean isDisposed = false;
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/SurveyPopUpDialog.java:55]: If let I code this, "= false" is unnecessary.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/CustomerSurveyHelper.java [hunk 7]
```diff
@@ -72,8 +71,6 @@ public void showFeedbackNotification(Project project) {
                     .subscribeOn(new IdeaSchedulers(project).dispatchUIThread())
                     .take(1)
                     .subscribe(next -> {
-                        SurveyPopUpDialog dialog = new SurveyPopUpDialog(CustomerSurveyHelper.this, project);
-                        dialog.setVisible(true);
                         synchronized (CustomerSurveyHelper.class) {
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/CustomerSurveyHelper.java:74]: lock class is not a good idea. although it is not changed by you in this PR.

Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/CustomerSurveyHelper.java:74]: This is the outdated codes which has been removed in next pr

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java (hunk 5 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java:67]: using the enum value(name) as survey type

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java (hunk 1 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java:40]: no, as get link differs for different survey, only sdk survey shares same strategy, i don't think this should be make as the default implementation

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java (hunk 4 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/CustomerSurvey.java:52]: super.getLink() will return the link value in constructor, which refers from constant.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/SurveyPopUpDialog.java [hunk 8]
```diff
@@ -2,21 +2,23 @@
...
 import com.intellij.ui.HyperlinkLabel;
+import com.intellij.ui.JBColor;
+import com.intellij.util.Consumer;
 import com.intellij.util.IconUtil;
 
-import com.microsoft.intellij.helpers.CustomerSurveyHelper;
-import com.microsoft.intellij.ui.util.UIUtils;
-import com.microsoft.intellij.util.PluginUtil;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.SystemUtils;
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-lib/src/main/java/com/microsoft/azure/toolkit/intellij/common/survey/SurveyPopUpDialog.java:19]: lang3.StringUtils ?

**Review by shenqianjin — APPROVED** [review]
LGTM

**Pull request merged**